cpp/mrc/src/public/coroutines/task_container.cpp

```diff
@@ -133,6 +140,19 @@ auto TaskContainer::gc_internal() -> std::size_t
     return deleted;
 }
 
+void TaskContainer::start_next_task()
+{
+    auto pos = [this]() {
+        std::scoped_lock lk{m_mutex};
+        auto pos = m_next_tasks.front();
+        m_next_tasks.pop();
+        return pos;
+    }();
+
+    // Start executing from the cleanup task to schedule the user's task onto the thread pool.
+    pos->value().resume();
```

Comment:
To be safe, we should add some sort of while check that can run "catch up". For example, if there is a weird unknown race condition which decrements `m_size` but does not launch a new task, you would end up with a situation where `m_size < (m_max_outstanding + 2)` and `m_next_tasks.size() == 2` (meaning 2 tasks could be started at once. A while loop could correct for that here.

Reply:
the TaskContainer is the only thing that can start and stop these tasks, so if there is such a race condition, that would be a bug. catch up logic would just hide the bug.